### DISCUSSION TITLE
Blueprint/Plotter life cycle

### DISCUSSION THREAD

**Discussion by shlomitk1** [fybrik/fybrik#556] [Developers]
Current behavior:
- M4DApplication owns Plotter. Plotter owns Blueprint(s)
- When M4DApplication is deleted, the owned Plotter is deleted as well, then the relevant Blueprints are deleted, including the installed resources.
- When M4DApplication is updated:
  - Per successful Plotter update: Blueprints are deleted/created/updated according to new requirements.
  - If Plotter generation fails, an error is returned. Existing Plotter/Blueprints are not affected.
  
The latter presents a problem. Plotter sends updates which are ignored if  M4DApplication is in an error state. The controller needs to distinguish between errors that come from Plotter and generation errors.
Approach1: delete an existing plotter if plotter re-generation failed. This is not efficient if there is a quick recovery and disrupts user work with the previous M4DApplication version (example: adding a new dataset fails, the user can still work with the old one).
Approach2: label generated Plotter version (in M4DApplication status) using the last observed version of M4DApplication that generated this Plotter. This guarantees that 1) earlier version of Plotter is still in use and 2) only relevant Plotter events are reported.  
@roee88 @ronenkat @froesef @simanadler

**Comment by simanadler** [comment]
@shlomitk1 This is a really important point you are raising.  Of the two options you suggested, I think approach2 makes more sense.

I think this issue is raises a larger question and is related to a request we recently heard about managing statuses (and of course virtual endpoints) at the data set level.  Continuing with that train of thought, should the plotters / blueprints also be for individual data sets?  @roee88 @ronenkat @froesef your thoughts?
